src/Raven.Client/Documents/Session/SessionInfo.cs

```diff
@@ -1,21 +1,73 @@
-﻿namespace Raven.Client.Documents.Session
+﻿using System;
+using System.Threading.Tasks;
+using Raven.Client.Http;
+using Sparrow;
+
+namespace Raven.Client.Documents.Session
 {
     public class SessionInfo
     {
-        public int? SessionId { get;}
+        [ThreadStatic]
+        private static int _clientSessionIdCounter;
+        protected int _clientSessionId;
+
+        public int SessionId { get;}
 
         public long? LastClusterTransactionIndex { get; set; }
 
         public bool AsyncCommandRunning { get; set; }
 
         public bool NoCaching { get; set; }
 
-        public SessionInfo(int sessionId, bool asyncCommandRunning, long? lastClusterTransactionIndex = null, bool noCaching = false)
+        public SessionInfo(string sessionContextKey, int writeBalanceSeed, bool asyncCommandRunning, long? lastClusterTransactionIndex = null, bool noCaching = false)
         {
+            SetSessionContext(sessionContextKey, writeBalanceSeed);
+
             LastClusterTransactionIndex = lastClusterTransactionIndex;
-            SessionId = sessionId;
             AsyncCommandRunning = asyncCommandRunning;
             NoCaching = noCaching;
         }
+
+        public void SetSessionContext(string sessionKey, int writeBalanceSeed = 0)
```

Comment:
The publicly available method should be IMO:
```
public void SetContext(string sessionKey)
```

and should check if sessionKey is not null or whitespace and use the original `loadBalanceSeed` from the conventions (the one passed via ctor).

Reply:
Can we have a test for this?

Reply:
I think we should also prevent user from setting this more than once in a lifecycle of a session. I do not see a valid scenario for this, it can only cause confusion and we should be safe by default.